src/client.rs

```diff
@@ -88,7 +153,106 @@ impl Record {
 
     // Return a field value.
     pub fn get(&self, key: &str) -> Option<&serde_json::Value> {
-        self.0.get(key)
+        self.value.get(key)
+    }
+
+    /// Return the attachment for this record, if any.
+    ///
+    /// Return values:
+    /// * `Some(Ok(_))` - There is an attachment, and it was successfully
+    ///   retrieved and converted to the requested type.
+    /// * `Some(Err(_))` - There is an attachment for this record, but it could
+    ///   not be retrieved or could not be converted to the requested type.
+    /// * `None` - There is no attachment for this record
+    ///
+    /// The `Some(Ok(_))` and `None` states are not expected to change. The
+    /// `Some(Err(_))` state may be temporary, and can be retried.
+    #[cfg(feature = "attachments")]
+    pub async fn get_attachment<T, E>(&mut self, client: &mut Client) -> Option<Result<T, ClientError>>
+    where T: TryFrom<Vec<u8>, Error = E>,
+        E: 'static + Send + Sync + std::error::Error
+    {
+        if let Attachment::Pending = self.attachment {
+            let maybe_attachment = self.value.as_object()
+                .and_then(|obj| obj.get("attachment"))
+                .map(|a| serde_json::from_value::<AttachmentMetadata>(a.clone()));
+
+            match maybe_attachment {
+                Some(Ok(meta)) => {
+                    self.attachment = Attachment::Some(meta);
```

Comment:
Can we parse the meta and make it available even when the attachment is in the pending state? Having the meta doesn't require us to fetch the attachment. Merino seem to expect this when populating "icons", see [here](https://github.com/mozilla-services/merino/blob/55d4819141658c4fef959db9f5ed616be5d50633/merino-adm/src/remote_settings/mod.rs#L141-L143)

Reply:
This is a good point. The reason for the delayed parsing is to because we don't control the creation of the object via serde. However, separating the methods that get the attachment data vs download it seem like a good idea.